artcommon/artcommonlib/konflux/konflux_db.py

```diff
@@ -4,27 +4,312 @@
+
+        :param group: Group name
+        :return: True if group is cached
+        """
+        with self._lock:
+            return group in self._groups
+
+    def stats(self, group: typing.Optional[str] = None) -> dict:
+        """
+        Get cache statistics.
+
+        :param group: Optional group to get stats for. If None, returns aggregate stats.
+        :return: Dictionary with cache stats
+        """
+        with self._lock:
+            total_queries = self._cache_hits + self._cache_misses
+            hit_rate = (self._cache_hits / total_queries * 100) if total_queries > 0 else 0
+
+            if group and group in self._groups:
+                # Group-specific stats
+                group_cache = self._groups[group]
+                return {
+                    'group': group,
+                    'total_builds': group_cache['total_builds'],
+                    'unique_names': len(group_cache['by_name']),
+                    'unique_nvrs': len(group_cache['by_nvr']),
+                    'oldest_build': group_cache['oldest'].isoformat() if group_cache['oldest'] else None,
+                    'newest_build': group_cache['newest'].isoformat() if group_cache['newest'] else None,
+                    'cache_hits': self._cache_hits,
+                    'cache_misses': self._cache_misses,
+                    'hit_rate': f"{hit_rate:.1f}%",
+                }
+            else:
+                # Aggregate stats across all groups
+                total_builds = sum(g['total_builds'] for g in self._groups.values())
+                total_nvrs = sum(len(g['by_nvr']) for g in self._groups.values())
+
+                return {
+                    'groups_cached': list(self._groups.keys()),
+                    'total_builds': total_builds,
+                    'unique_nvrs': total_nvrs,
+                    'cache_hits': self._cache_hits,
+                    'cache_misses': self._cache_misses,
+                    'hit_rate': f"{hit_rate:.1f}%",
+                }
+
+    def clear(self, group: typing.Optional[str] = None):
+        """
+        Clear cached data.
+
+        :param group: Optional group to clear. If None, clears all groups.
+        """
+        with self._lock:
+            if group:
+                if group in self._groups:
+                    del self._groups[group]
+                    self.logger.info(f"Cache cleared for group '{group}'")
+            else:
+                self._groups.clear()
+                self._cache_hits = 0
+                self._cache_misses = 0
+                self.logger.info("Cache cleared for all groups")
 
 
 class KonfluxDb:
-    def __init__(self):
+    # Class-level singleton cache shared across all KonfluxDb instances
+    _shared_cache: typing.Optional[BuildCache] = None
+    _cache_lock = threading.RLock()
+    _group_loading_events: typing.Dict[str, asyncio.Event] = {}  # Per-group events for coordinating lazy-load
```

Comment:
I haven't looked at this closely, but it might worth mentioning that `asyncio.Event` is not thread-safe. Could you explain more about how those events are used?

Reply:
I'll take a closer look after my release artistry work this week if this PR is still open.